existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,264 @@
+```
+There are a few things to mention here.
+
+* In the type signature, `a` and `c` are universally quantified variables, and `b` is existentially quantified. This is what I was talking about when I said that type variables may belong to different quantifiers.
+* The same variable can't be quantified more than once.
+* You can see that constraints for `b` are added in the middle of the declaration of the function type (because `b` was introduced in the middle of the function).
+* Now you can't manually choose type for `b` – the compiler must deduce the correct variable you want to use by itself.
+
+Here's a real-life example of using existential types in functions.
+Imagine a function that prints some logs while calculating the result.
+You may want to write logs into a console or into some file.
+In order to do it, you pass a logging function as an argument.
+
+```haskell
+func :: _ -> IO ()
+func log = do
+  username <- getUsername
+  log username
+
+  userCount <- getUserCount
+  log userCount
+
+getUsername :: IO String
+getUsername = undefined
+
+getUserCount :: IO Int
+getUserCount = undefined
+```
+
+What type should the log function have?
+Since it somehow logs both `String` and `Int`:
+
+```haskell
+log :: Show a => a -> IO ()
+```
+However, if we add this signature to `func` and try to call it with different arguments:
+
+```haskell
+func :: Show a => (a -> IO ()) -> IO ()
+func log = ...
+
+main :: IO ()
+main = do
+  -- log to the console
+  func print
+
+  -- log to a logfile
+  func (appendFile "logfile.log" . show)
+```
+We will see some nasty compiler errors:
+```
+Ambiguous type variable ‘a1’ arising from a use of ‘func’
+```
+
+This is because the compiler can't deduce the type of `a` from the given constraints.
+If we use `forall.` explicitly as we did before, it wants help, because in this case, the caller is the one who should specify the type of the log function. But we can't specify it because it is different for different logging operations.
+The only solution here is to add constraints and `forall.` quantifier to the log function signature:
+
+```haskell
+{-# LANGUAGE RankNTypes #-}
+
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+Now we put all responsibility for deducing types on the shoulders of the compiler, and it will deduce them the way it wants.
+
+#### Hiding type variables
+
+```Haskell
+data MyData a = forall b. MkMyData b a
+-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
+```
+
+The first thing you might have noticed is that we don't have a type variable for `b` on the left side of the expression.
+However, we still can use it on the right side.
+This is because `b` is existentially quantified.
+The compiler will deduce the type of the variable itself, and you can't specify it.
+
+```haskell
+func :: MyData Integer -> IO ()
+func (MkMyData (x :: Double) y) = print x -- compiler error - can't match type of x with Double
```

Comment:
Ok, but how do I use it meaningfully?

Also, an interesting take: to me it looks like co-phantom data type. :)

Reply:
Then maybe consider replacing `MyData` with something that can be meaningfully used.

Remember: you're trying to convince the reader that these things _are useful_ and therefore they should pay close attention to your explanation.

When you use examples that are not meaningful or are too farfetched/contrived, the reader may begin scratching their head and stop reading.

Reply:
I moved this example to SomeException section, with explaining that you need to use special function to extract exception